File: src/vehicle/vehicleClient.ts
import { useEffect, useState } from "react";
import type { VehicleState } from "../shared/vehicleTypes";

type Listener = (state: VehicleState) => void;

type Track = Omit<VehicleState["audio"]["nowPlaying"], "positionSec" | "isPlaying">;

const tracks: Track[] = [
  {
    title: "Midnight Lights",
    artist: "Eliora",
    album: "Signals",
    durationSec: 256,
  },
  {
    title: "Neon Drift",
    artist: "Atlas Run",
    album: "Night City",
    durationSec: 212,
  },
  {
    title: "Magnetic",
    artist: "Nova Phase",
    album: "Binary Sky",
    durationSec: 284,
  },
];

let trackIndex = 0;

let rawState: VehicleState = {
  turn: {
    mode: "off",
    left: false,
    right: false,
  },
  engine: {
    rpm: 1650,
  },
  vehicle: {
    speedKmh: 54,
  },
  fuel: {
    percent: 47,
  },
  temp: {
    oilC: 82,
    coolantC: 80,
  },
  electrical: {
    batteryV: 14.4,
  },
  climate: {
    tempSetC: 21,
    fan: 2,
    ac: true,
    recirc: false,
    defrost: false,
    auto: true,
  },
  audio: {
    volume: 38,
    muted: false,
    source: "bt",
    nowPlaying: {
      ...tracks[0],
      positionSec: 114,
      isPlaying: true,
    },
  },
  car: {
    lights: false,
    hazards: false,
    locked: true,
  },
  ambient: {
    color: "#7EE3FF",
    brightness: 65,
  },
};

let state: VehicleState = rawState;
const listeners = new Set<Listener>();
let socket: WebSocket | null = null;
let ytmSocket: WebSocket | null = null;
let connectUrl = `ws://${window.location.hostname}:8765`;
let ytmOverride: Partial<VehicleState["audio"]["nowPlaying"]> | null = null;
let btOverride: Partial<VehicleState["audio"]["nowPlaying"]> | null = null;
let btPollTimer: number | null = null;
const btBaseUrl = `http://${window.location.hostname}:5175`;
const ytmUrls = (() => {
  const host = window.location.hostname;
  const urls = [`ws://${host}:5174`];
  if (host !== "localhost") {
    urls.push("ws://localhost:5174");
  }
  return urls;
})();
let ytmUrlIndex = 0;

const BLINK_INTERVAL_MS = 330;
let blinkTimer: number | null = null;
let blinkPhaseOn = false;
let blinkEnabled = false;
let lastBlinkApplied = false;

const MOCK_INTERVAL_MS = 120;
let mockTimer: number | null = null;
let mockTick = 0;

const notify = () => {
  listeners.forEach((listener) => listener(state));
};

const commit = (next: VehicleState) => {
  rawState = next;
  deriveState();
};

const deriveState = () => {
  const hazardsOn = rawState.car.hazards;
  const mode = hazardsOn ? "hazard" : rawState.turn.mode;
  blinkEnabled = hazardsOn || mode !== "off";
  if (!blinkEnabled) {
    blinkPhaseOn = false;
    state = {
      ...rawState,
      turn: {
        ...rawState.turn,
        mode,
        left: rawState.turn.left,
        right: rawState.turn.right,
      },
    };
    return;
  }
  const leftOn = blinkPhaseOn && (mode === "left" || mode === "hazard");
  const rightOn = blinkPhaseOn && (mode === "right" || mode === "hazard");
  state = {
    ...rawState,
    turn: {
      ...rawState.turn,
      mode,
      left: leftOn,
      right: rightOn,
    },
  };
};

const applyBlinkPhase = () => {
  const shouldApply = blinkEnabled && blinkPhaseOn;
  if (shouldApply === lastBlinkApplied && blinkEnabled === false) return;
  lastBlinkApplied = shouldApply;
  deriveState();
  notify();
};

const startBlinkLoop = () => {
  if (blinkTimer !== null) return;
  blinkTimer = window.setInterval(() => {
    if (!blinkEnabled) {
      if (lastBlinkApplied) {
        lastBlinkApplied = false;
        blinkPhaseOn = false;
        deriveState();
        notify();
      }
      return;
    }
    blinkPhaseOn = !blinkPhaseOn;
    applyBlinkPhase();
  }, BLINK_INTERVAL_MS);
};

export const connectVehicleWS = (url: string) => {
  connectUrl = url;
  connect();
};

const connect = () => {
  if (socket && socket.readyState === WebSocket.OPEN) return;

  try {
    socket = new WebSocket(connectUrl);
  } catch {
    return;
  }

  socket.addEventListener("open", () => {
    stopMockLoop();
  });

  socket.addEventListener("message", (event) => {
    try {
      const message = JSON.parse(event.data as string) as {
        type: string;
        payload: VehicleState | (Omit<VehicleState, "ambient"> & { ambient?: VehicleState["ambient"] });
      };
      if (message.type === "state" && message.payload) {
        rawState = {
          ...(message.payload as VehicleState),
          ambient: message.payload.ambient ?? rawState.ambient,
        };
        if (btOverride) {
          rawState = {
            ...rawState,
            audio: {
              ...rawState.audio,
              nowPlaying: {
                ...rawState.audio.nowPlaying,
                ...btOverride,
              },
            },
          };
        }
        if (ytmOverride) {
          rawState = {
            ...rawState,
            audio: {
              ...rawState.audio,
              nowPlaying: {
                ...rawState.audio.nowPlaying,
                ...ytmOverride,
              },
            },
          };
        }
        deriveState();
        notify();
      }
    } catch {
      // Ignore malformed messages
    }
  });

  socket.addEventListener("close", () => {
    socket = null;
    setTimeout(connect, 2000);
  });

  socket.addEventListener("error", () => {
    // stay idle without mock updates
  });
};

const applyYtmNowPlaying = (payload: {
  title?: string;
  artist?: string;
  album?: string;
  artworkUrl?: string;
  durationSec?: number;
  positionSec?: number;
  isPlaying?: boolean;
}) => {
  if (!payload?.title && !payload?.artist) return;
  ytmOverride = {
    title: payload.title ?? rawState.audio.nowPlaying.title,
    artist: payload.artist ?? rawState.audio.nowPlaying.artist,
    album: payload.album ?? rawState.audio.nowPlaying.album,
    artworkUrl: payload.artworkUrl ?? rawState.audio.nowPlaying.artworkUrl,
    durationSec: Number.isFinite(payload.durationSec)
      ? Number(payload.durationSec)
      : rawState.audio.nowPlaying.durationSec,
    positionSec: Number.isFinite(payload.positionSec)
      ? Number(payload.positionSec)
      : rawState.audio.nowPlaying.positionSec,
    isPlaying:
      typeof payload.isPlaying === "boolean"
        ? payload.isPlaying
        : rawState.audio.nowPlaying.isPlaying,
  };
  commit({
    ...rawState,
    audio: {
      ...rawState.audio,
      nowPlaying: {
        ...rawState.audio.nowPlaying,
        ...ytmOverride,
      },
    },
  });
  notify();
};

const applyBtNowPlaying = (payload: {
  connected?: boolean;
  title?: string;
  artist?: string;
  album?: string;
  durationSec?: number;
  positionSec?: number;
  isPlaying?: boolean;
}) => {
  if (!payload?.connected) {
    btOverride = null;
    return;
  }
  btOverride = {
    title: payload.title ?? rawState.audio.nowPlaying.title,
    artist: payload.artist ?? rawState.audio.nowPlaying.artist,
    album: payload.album ?? rawState.audio.nowPlaying.album,
    durationSec: Number.isFinite(payload.durationSec)
      ? Number(payload.durationSec)
      : rawState.audio.nowPlaying.durationSec,
    positionSec: Number.isFinite(payload.positionSec)
      ? Number(payload.positionSec)
      : rawState.audio.nowPlaying.positionSec,
    isPlaying:
      typeof payload.isPlaying === "boolean"
        ? payload.isPlaying
        : rawState.audio.nowPlaying.isPlaying,
  };
  commit({
    ...rawState,
    audio: {
      ...rawState.audio,
      nowPlaying: {
        ...rawState.audio.nowPlaying,
        ...btOverride,
      },
    },
  });
  notify();
};

const connectYtm = () => {
  if (ytmSocket && (ytmSocket.readyState === WebSocket.OPEN || ytmSocket.readyState === WebSocket.CONNECTING)) {
    return;
  }

  const url = ytmUrls[ytmUrlIndex % ytmUrls.length];
  ytmUrlIndex += 1;

  try {
    ytmSocket = new WebSocket(url);
  } catch {
    ytmSocket = null;
    setTimeout(connectYtm, 2000);
    return;
  }

  ytmSocket.addEventListener("message", (event) => {
    try {
      const payload = JSON.parse(event.data as string);
      applyYtmNowPlaying(payload);
    } catch {
      // ignore malformed
    }
  });

  ytmSocket.addEventListener("close", () => {
    ytmSocket = null;
    setTimeout(connectYtm, 2000);
  });

  ytmSocket.addEventListener("error", () => {
    // reconnect on close
  });
};

const startBluetoothNowPlayingLoop = () => {
  if (btPollTimer !== null) return;
  btPollTimer = window.setInterval(async () => {
    try {
      const res = await fetch(`${btBaseUrl}/media/now-playing`);
      if (!res.ok) return;
      const payload = (await res.json()) as {
        connected?: boolean;
        title?: string;
        artist?: string;
        album?: string;
        durationSec?: number;
        positionSec?: number;
        isPlaying?: boolean;
      };
      if (ytmOverride) return;
      applyBtNowPlaying(payload);
    } catch {
      // ignore
    }
  }, 3000);
};

export const sendVehicleCommand = (type: string, payload?: unknown) => {
  if (socket && socket.readyState === WebSocket.OPEN) {
    socket.send(JSON.stringify({ type, payload }));
    applyCommand(type, payload);
    notify();
    return;
  }

  applyCommand(type, payload);
  notify();
};

const clamp = (value: number, min: number, max: number) => Math.min(max, Math.max(min, value));

const setTrack = (direction: 1 | -1) => {
  trackIndex = (trackIndex + direction + tracks.length) % tracks.length;
  const next = tracks[trackIndex];
  commit({
    ...rawState,
    audio: {
      ...rawState.audio,
      nowPlaying: {
        ...next,
        positionSec: 0,
        isPlaying: true,
      },
    },
  });
};

const applyCommand = (type: string, payload?: unknown) => {
  switch (type) {
    case "climate/set": {
      const next = payload as Partial<VehicleState["climate"]>;
      commit({
        ...rawState,
        climate: {
          ...rawState.climate,
          ...next,
          tempSetC: next?.tempSetC !== undefined ? clamp(next.tempSetC, 16, 28) : rawState.climate.tempSetC,
          fan: next?.fan !== undefined ? clamp(next.fan, 0, 5) : rawState.climate.fan,
        },
      });
      return;
    }
    case "audio/set": {
      const next = payload as Partial<VehicleState["audio"]>;
      commit({
        ...rawState,
        audio: {
          ...rawState.audio,
          ...next,
        },
      });
      return;
    }
    case "audio/control": {
      const action = (payload as { action?: string })?.action;
      if (action === "toggle") {
        commit({
          ...rawState,
          audio: {
            ...rawState.audio,
            nowPlaying: {
              ...rawState.audio.nowPlaying,
              isPlaying: !rawState.audio.nowPlaying.isPlaying,
            },
          },
        });
      }
      if (action === "play") {
        commit({
          ...rawState,
          audio: {
            ...rawState.audio,
            nowPlaying: {
              ...rawState.audio.nowPlaying,
              isPlaying: true,
            },
          },
        });
      }
      if (action === "pause") {
        commit({
          ...rawState,
          audio: {
            ...rawState.audio,
            nowPlaying: {
              ...rawState.audio.nowPlaying,
              isPlaying: false,
            },
          },
        });
      }
      if (action === "next") {
        setTrack(1);
      }
      if (action === "prev") {
        setTrack(-1);
      }
      return;
    }
    case "turn/set":
      commit({
        ...rawState,
        turn: {
          ...rawState.turn,
          ...(payload as Partial<VehicleState["turn"]>),
        },
      });
      return;
    case "car/toggleLights":
      commit({ ...rawState, car: { ...rawState.car, lights: !rawState.car.lights } });
      return;
    case "car/toggleHazards": {
      const hazardsOn = !rawState.car.hazards;
      commit({
        ...rawState,
        car: { ...rawState.car, hazards: hazardsOn },
        turn: {
          ...rawState.turn,
          mode: hazardsOn ? "hazard" : "off",
          left: false,
          right: false,
        },
      });
      return;
    }
    case "car/toggleLock":
      commit({ ...rawState, car: { ...rawState.car, locked: !rawState.car.locked } });
      return;
    case "ambient/set": {
      const next = payload as Partial<VehicleState["ambient"]>;
      commit({
        ...rawState,
        ambient: {
          ...rawState.ambient,
          ...next,
        },
      });
      return;
    }
    default:
      return;
  }
};

const startMockLoop = () => {
  if (mockTimer !== null) return;
  mockTimer = window.setInterval(() => {
    if (socket && socket.readyState === WebSocket.OPEN) {
      stopMockLoop();
      return;
    }
    mockTick += 0.03;
    const rpm = 800 + ((Math.sin(mockTick) + 1) / 2) * 7200;
    const speedKmh = ((Math.sin(mockTick * 0.7 + 1.2) + 1) / 2) * 200;
    commit({
      ...rawState,
      engine: {
        ...rawState.engine,
        rpm,
      },
      vehicle: {
        ...rawState.vehicle,
        speedKmh,
      },
    });
    notify();
  }, MOCK_INTERVAL_MS);
};

const stopMockLoop = () => {
  if (mockTimer === null) return;
  window.clearInterval(mockTimer);
  mockTimer = null;
};

export const useVehicleState = () => {
  const [vehicleState, setVehicleState] = useState(state);

  useEffect(() => {
    connect();
    connectYtm();
    startBluetoothNowPlayingLoop();
    startBlinkLoop();
    startMockLoop();
    const unsubscribe = subscribe(setVehicleState);
    return () => unsubscribe();
  }, []);

  return vehicleState;
};

export const subscribe = (listener: Listener) => {
  listeners.add(listener);
  deriveState();
  listener(state);
  return () => {
    listeners.delete(listener);
  };
};

File: qt-dash/src/VehicleClient.cpp
#include "VehicleClient.h"

#include <QJsonDocument>
#include <QtMath>

VehicleClient::VehicleClient(QObject *parent)
    : QObject(parent),
      m_state(defaultState()) {
    m_reconnectTimer.setInterval(2000);
    m_reconnectTimer.setSingleShot(true);
    connect(&m_reconnectTimer, &QTimer::timeout, this, &VehicleClient::reconnect);

    m_mockTimer.setInterval(120);
    connect(&m_mockTimer, &QTimer::timeout, this, &VehicleClient::advanceMock);

    connect(&m_socket, &QWebSocket::connected, this, &VehicleClient::handleConnected);
    connect(&m_socket, &QWebSocket::disconnected, this, &VehicleClient::handleDisconnected);
    connect(&m_socket, &QWebSocket::textMessageReceived, this, &VehicleClient::handleMessage);

    startMock();
}

QJsonObject VehicleClient::state() const {
    return m_state;
}

bool VehicleClient::connected() const {
    return m_connected;
}

void VehicleClient::connectTo(const QUrl &url) {
    m_url = url;
    if (m_simulationOnly) {
        startMock();
        return;
    }
    reconnect();
}

void VehicleClient::sendCommand(const QString &type, const QJsonObject &payload) {
    if (m_socket.state() != QAbstractSocket::ConnectedState) {
        if (type == QStringLiteral("bt/media/control")) {
            const QString action = payload.value(QStringLiteral("action")).toString();
            QJsonObject next = m_state;
            QJsonObject audio = next.value(QStringLiteral("audio")).toObject();
            QJsonObject nowPlaying = audio.value(QStringLiteral("nowPlaying")).toObject();

            if (action == QStringLiteral("play")) {
                nowPlaying.insert(QStringLiteral("isPlaying"), true);
            } else if (action == QStringLiteral("pause")) {
                nowPlaying.insert(QStringLiteral("isPlaying"), false);
            }

            audio.insert(QStringLiteral("nowPlaying"), nowPlaying);
            next.insert(QStringLiteral("audio"), audio);
            setState(next);
        }
        return;
    }

    QJsonObject message;
    message.insert(QStringLiteral("type"), type);
    message.insert(QStringLiteral("payload"), payload);
    m_socket.sendTextMessage(QString::fromUtf8(QJsonDocument(message).toJson(QJsonDocument::Compact)));
}

void VehicleClient::setSimulationOnly(bool simulationOnly) {
    if (m_simulationOnly == simulationOnly) {
        return;
    }

    m_simulationOnly = simulationOnly;
    if (m_simulationOnly) {
        m_reconnectTimer.stop();
        m_socket.close();
        setConnected(false);
        startMock();
    } else {
        reconnect();
    }
}

void VehicleClient::handleConnected() {
    stopMock();
    setConnected(true);
}

void VehicleClient::handleDisconnected() {
    setConnected(false);
    startMock();
    m_reconnectTimer.start();
}

void VehicleClient::handleMessage(const QString &message) {
    const QJsonDocument document = QJsonDocument::fromJson(message.toUtf8());
    if (!document.isObject()) {
        return;
    }

    const QJsonObject object = document.object();
    if (object.value(QStringLiteral("type")).toString() != QStringLiteral("state")) {
        return;
    }

    const QJsonObject payload = object.value(QStringLiteral("payload")).toObject();
    if (!payload.isEmpty()) {
        setState(payload);
    }
}

void VehicleClient::reconnect() {
    if (m_simulationOnly || !m_url.isValid() || m_socket.state() == QAbstractSocket::ConnectedState ||
        m_socket.state() == QAbstractSocket::ConnectingState) {
        return;
    }

    m_socket.open(m_url);
}

void VehicleClient::advanceMock() {
    m_mockTick += 0.03;
    QJsonObject next = m_state;

    QJsonObject engine = next.value(QStringLiteral("engine")).toObject();
    engine.insert(QStringLiteral("rpm"), 800 + ((qSin(m_mockTick) + 1) / 2) * 7200);
    next.insert(QStringLiteral("engine"), engine);

    QJsonObject vehicle = next.value(QStringLiteral("vehicle")).toObject();
    vehicle.insert(QStringLiteral("speedKmh"), ((qSin(m_mockTick * 0.7 + 1.2) + 1) / 2) * 180);
    next.insert(QStringLiteral("vehicle"), vehicle);

    QJsonObject audio = next.value(QStringLiteral("audio")).toObject();
    QJsonObject nowPlaying = audio.value(QStringLiteral("nowPlaying")).toObject();
    const int duration = nowPlaying.value(QStringLiteral("durationSec")).toInt(0);
    if (nowPlaying.value(QStringLiteral("isPlaying")).toBool(false) && duration > 0) {
        m_mockMusicAccumulator += 0.12;
        int position = nowPlaying.value(QStringLiteral("positionSec")).toInt(0);
        while (m_mockMusicAccumulator >= 1.0) {
            position = (position + 1) % duration;
            m_mockMusicAccumulator -= 1.0;
        }
        nowPlaying.insert(QStringLiteral("positionSec"), position);
        audio.insert(QStringLiteral("nowPlaying"), nowPlaying);
        next.insert(QStringLiteral("audio"), audio);
    }

    setState(next);
}

void VehicleClient::setConnected(bool connected) {
    if (m_connected == connected) {
        return;
    }
    m_connected = connected;
    emit connectedChanged();
}

void VehicleClient::setState(const QJsonObject &state) {
    m_state = state;
    emit stateChanged();
}

void VehicleClient::startMock() {
    if (!m_mockTimer.isActive()) {
        m_mockTimer.start();
    }
}

void VehicleClient::stopMock() {
    if (m_mockTimer.isActive()) {
        m_mockTimer.stop();
    }
}

QJsonObject VehicleClient::defaultState() const {
    return QJsonObject{
        {QStringLiteral("turn"),
         QJsonObject{{QStringLiteral("mode"), QStringLiteral("off")},
                     {QStringLiteral("left"), false},
                     {QStringLiteral("right"), false}}},
        {QStringLiteral("engine"), QJsonObject{{QStringLiteral("rpm"), 1650}}},
        {QStringLiteral("vehicle"), QJsonObject{{QStringLiteral("speedKmh"), 54}}},
        {QStringLiteral("fuel"), QJsonObject{{QStringLiteral("percent"), 47}}},
        {QStringLiteral("temp"),
         QJsonObject{{QStringLiteral("oilC"), 82}, {QStringLiteral("coolantC"), 80}}},
        {QStringLiteral("electrical"), QJsonObject{{QStringLiteral("batteryV"), 14.4}}},
        {QStringLiteral("climate"),
         QJsonObject{{QStringLiteral("tempSetC"), 21},
                     {QStringLiteral("fan"), 2},
                     {QStringLiteral("ac"), true},
                     {QStringLiteral("recirc"), false},
                     {QStringLiteral("defrost"), false},
                     {QStringLiteral("auto"), true}}},
        {QStringLiteral("audio"),
         QJsonObject{{QStringLiteral("volume"), 38},
                     {QStringLiteral("muted"), false},
                     {QStringLiteral("source"), QStringLiteral("bt")},
                     {QStringLiteral("nowPlaying"),
                      QJsonObject{{QStringLiteral("title"), QStringLiteral("Midnight Lights")},
                                  {QStringLiteral("artist"), QStringLiteral("Eliora")},
                                  {QStringLiteral("album"), QStringLiteral("Signals")},
                                  {QStringLiteral("artworkUrl"),
                                   QStringLiteral("file:///home/admin/digital-dash/public/albumcover.jpg")},
                                  {QStringLiteral("durationSec"), 256},
                                  {QStringLiteral("positionSec"), 114},
                                  {QStringLiteral("isPlaying"), true}}}}},
        {QStringLiteral("car"),
         QJsonObject{{QStringLiteral("lights"), false},
                     {QStringLiteral("hazards"), false},
                     {QStringLiteral("locked"), true}}},
        {QStringLiteral("ambient"),
         QJsonObject{{QStringLiteral("color"), QStringLiteral("#7EE3FF")},
                     {QStringLiteral("brightness"), 65}}},
    };
}
